astropy/tests/helper.py

```diff
@@ -192,11 +194,72 @@ def assert_quantity_allclose(actual, desired, rtol=1.0e-7, atol=None, **kwargs):
     This is a :class:`~astropy.units.Quantity`-aware version of
     :func:`numpy.testing.assert_allclose`.
     """
-    import numpy as np
-
     from astropy.units.quantity import _unquantify_allclose_arguments
 
     __tracebackhide__ = True
     np.testing.assert_allclose(
         *_unquantify_allclose_arguments(actual, desired, rtol, atol), **kwargs
     )
+
+
+def assert_objects_equal(actual, desired, /, *, attrs=None, compare_class=True):
+    """Compare objects by their shape and attributes, including those provided by .info
+
+    Parameters
+    ----------
+    actual and desired: np.ndarray (positional only)
```

Comment:
They don't have to be `ndarray` instances (sorry i missed this earlier). Maybe just `object`? Though they do need to have `.info`.

Reply:
they also need a `shape` attribute, so this really feels like it was only designed for `ndarray` and subclasses to me (even though the function name implies a broader application). Furthermore, in typed Python, `object` designates the type that has *no* expected methods or attribute, so this would be a source of confusion.

Reply:
We can ask for `~astropy.utils.shapes.ShapedLikeNDArray`?

Reply:
hum... the proper way to define the expected type would then seem to be a `typing.Protocol`, but perhaps a simpler way to go about it is to say `ndarray or ShapedLikeNDArray` ?

Reply:
In astropy, I think `ndarary subclass or ShapedLikeNDArray` would cover it. Really, they also need to carry `.info` (which excludes `ndarray` proper).

Reply:
Thanks ! given how non-obvious it was to me, I actually added the full protocol hints, which enables IDEs to check this function is used correctly. Let me know what you think.